core/shared/src/main/scala/fs2/text.scala

```diff
@@ -35,74 +38,119 @@ object text {
      * 0 is returned.
      */
     def lastIncompleteBytes(bs: Array[Byte]): Int = {
-      val lastThree = bs.drop(0.max(bs.size - 3)).toArray.reverseIterator
-      lastThree
-        .map(continuationBytes)
-        .zipWithIndex
-        .find {
-          case (c, _) => c >= 0
-        }
-        .map {
-          case (c, i) => if (c == i) 0 else i + 1
+      /*
+       * This is logically the same as this
+       * code, but written in a low level way
+       * to avoid any allocations and just do array
+       * access
+       *
+       *
+       *
+        val lastThree = bs.drop(0.max(bs.size - 3)).toArray.reverseIterator
+        lastThree
+          .map(continuationBytes)
+          .zipWithIndex
+          .find {
+            case (c, _) => c >= 0
+          }
+          .map {
+            case (c, i) => if (c == i) 0 else i + 1
+          }
+          .getOrElse(0)
+
+       */
+
+      val minIdx = 0.max(bs.size - 3)
+      var idx = bs.size - 1
+      var counter = 0
+      var res = 0
+      while (minIdx <= idx) {
+        val c = continuationBytes(bs(idx))
+        if (c >= 0) {
+          if (c == counter)
+            res = 0
+          else
+            res = counter + 1
```

Comment:
I'm pretty sure `res` is already `0` so i think you just want `if (c != counter) res = counter + 1` don't you?